discovery/manager_test.go

```diff
@@ -0,0 +1,726 @@
+						}
+
+						t.Errorf("%v. %v: \ntargets mismatch \nreceived: %v \nexpected: %v",
+							x, testCase.title,
+							receivedFormated,
+							expectedFormated)
+					}
+				}
+			}
+		}
+	}
+}
+
+func TestTargetSetRecreatesTargetGroupsEveryRun(t *testing.T) {
+	verifyPresence := func(tSets map[string]map[string][]*config.TargetGroup, tSetName string, provName, label string, present bool) {
+		if _, ok := tSets[tSetName]; !ok {
+			t.Fatalf("'%s' should be present in TargetSets: %v", tSetName, tSets)
+			return
+		}
+		if _, ok := tSets[tSetName][provName]; !ok {
+			t.Fatalf("'%s' should be present in Discovery providers: %v", provName, tSets[tSetName])
+			return
+		}
+
+		match := false
+		var mergedTargets string
+		for _, targetGroup := range tSets[tSetName][provName] {
+
+			for _, l := range targetGroup.Targets {
+				mergedTargets = mergedTargets + " " + l.String()
+				if l.String() == label {
+					match = true
+				}
+			}
+
+		}
+		if match != present {
+			msg := ""
+			if !present {
+				msg = "not"
+			}
+			t.Fatalf("'%s' should %s be present in Targets labels: %v", label, msg, mergedTargets)
+		}
+	}
+
+	cfg := &config.Config{}
+
+	sOne := `
+scrape_configs:
+ - job_name: 'prometheus'
+   static_configs:
+   - targets: ["foo:9090"]
+   - targets: ["bar:9090"]
+`
+	if err := yaml.Unmarshal([]byte(sOne), cfg); err != nil {
+		t.Fatalf("Unable to load YAML config sOne: %s", err)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	discoveryManager := NewManager(ctx, nil)
+	go discoveryManager.Run()
+
+	discoveryManager.ApplyConfig(cfg)
+
+	_ = <-discoveryManager.SyncCh()
+	verifyPresence(discoveryManager.targets, "prometheus", "static/0", "{__address__=\"foo:9090\"}", true)
+	verifyPresence(discoveryManager.targets, "prometheus", "static/0", "{__address__=\"bar:9090\"}", true)
+
+	sTwo := `
+scrape_configs:
+ - job_name: 'prometheus'
+   static_configs:
+   - targets: ["foo:9090"]
+`
+	if err := yaml.Unmarshal([]byte(sTwo), cfg); err != nil {
+		t.Fatalf("Unable to load YAML config sOne: %s", err)
+	}
+	discoveryManager.ApplyConfig(cfg)
+
+	_ = <-discoveryManager.SyncCh()
+	verifyPresence(discoveryManager.targets, "prometheus", "static/0", "{__address__=\"foo:9090\"}", true)
+	verifyPresence(discoveryManager.targets, "prometheus", "static/0", "{__address__=\"bar:9090\"}", false)
+}
+
+type update struct {
+	targetGroups []config.TargetGroup
+	interval     time.Duration
+}
+
+type mockdiscoveryProvider struct {
+	updates []update
+	up      chan<- []*config.TargetGroup
+}
+
+func newMockDiscoveryProvider(updates []update) mockdiscoveryProvider {
+
+	tp := mockdiscoveryProvider{
+		updates: updates,
+	}
+	return tp
+}
+
+func (tp mockdiscoveryProvider) Run(ctx context.Context, up chan<- []*config.TargetGroup) {
+	tp.up = up
+	tp.sendUpdates()
+}
+
+func (tp mockdiscoveryProvider) sendUpdates() {
+	for _, update := range tp.updates {
+
+		time.Sleep(update.interval * time.Millisecond)
```

Comment:
Where do you suspect the code to be timing dependent to have the tests play with sleeps?
Also `update.interval` already is a duration but the test cases just add plain numbers, which would be nanoseconds. A proper duration should be specified there rather than implicitly multiplying it with milliseconds further down.

Reply:
Without sleeps you can't predict the order of the updates so you can't match with the expected result for the intermediate updates.  
The original discovery test was like this so I thought it covered some bug for the intermediate updates. If that is not the case than I can remove the sleeps and just match the final result to the expected result.